### PR TITLE
Issue #24 create cards endpoint

### LINKED ISSUE #24: Add endpoints to list categories and cards in a category
- A logged in user should be able to get a paged list of all categories available to them: `/api/categories?page=2`
- A logged in user should be able to get a paged list of all cards in a specified category: `/api/cards?category=xxxx&page=2`

### PR DESCRIPTION
**Pull request by alfabravo2013**
Currently, the `/api/cards` endpoint returns documents instead of their proper projections. This is a temporary solution, because we need to decide which information the server should return for each flashcard, and also find the best way to map data fetched from the DB.

### CODE REVIEW

File: server/src/main/java/org/hyperskill/community/flashcards/card/CardService.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+public class CardService {
+    private static final int pageSize = 20;
+    private final MongoTemplate mongoTemplate;
+    private final CategoryService categoryService;
+
+    public Page<Card> getCardsByCategory(String categoryId, int page) {
+        Objects.requireNonNull(categoryId, "Category ID cannot be null");
+        if (page < 0) {
+            throw new IllegalArgumentException("Page number is out of bounds");
+        }
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/CardService.java]: we could have hibernate validate on negative page param

File: server/src/main/java/org/hyperskill/community/flashcards/card/mapper/CardReadConverter.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            case "MCQ" -> MultipleChoiceQuiz.class;
+            case "QAC" -> QuestionAndAnswerCard.class;
+            default -> throw new IllegalStateException("Unknown value " + typeAlias);
+        };
+        try {
+            return objectMapper.readValue(source.toJson(), clazz);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/mapper/CardReadConverter.java:33]: wow - that looks real low-level :-)

Comment (wisskirchenj) [reply server/src/main/java/org/hyperskill/community/flashcards/card/mapper/CardReadConverter.java:33]: including the converter classes to come.

File: server/src/main/java/org/hyperskill/community/flashcards/category/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    private final CategoryService categoryService;
+
+    @GetMapping
+    public CategoryPageResponse getAllCategories(
+            @RequestParam(name = "page", required = false, defaultValue = "0") int page,
+            Authentication authentication
+    ) {
+        // fixme
+        var username = "test1@test.com";  // authentication.getName();
+        var categories = categoryService.getCategories(page);
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/category/CategoryController.java]: you can use @AuthenticationPrincipal Jwt jwt
and get the Jwt authentication object. Then getSubject() provides the requesters name

File: server/src/main/java/org/hyperskill/community/flashcards/category/CategoryService.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        var username = SecurityContextHolder.getContext().getAuthentication().getName();
+        if (username == null) {
+            log.debug("Principal's name from SecurityContext is null");
+            throw new AuthenticationCredentialsNotFoundException("Authentication is required");
+        }
+
+        // fixme change to actual username
+        var usernameHasReadPermission = Criteria.where("username").is("test1@test.com").and("permission").regex("(.*r.*){1,3}");
+        var currentUserHasAccess = Criteria.where("access").elemMatch(usernameHasReadPermission);
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/category/CategoryService.java:31]: the part above I don't get. Spring security ensures, that this method is called only authenticated.. and you can pass the username to the service

**Review by wisskirchenj — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/hyperskill/community/flashcards/card/mapper/CardReadConverter.java (hunk 2 above)
Comment (wisskirchenj) [reply server/src/main/java/org/hyperskill/community/flashcards/card/mapper/CardReadConverter.java:33]: no change needed here - just spontaneous comment :-)

File: server/src/main/java/org/hyperskill/community/flashcards/category/CategoryService.java (hunk 4 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/category/CategoryService.java:31]: fixed

File: server/src/main/java/org/hyperskill/community/flashcards/category/CategoryController.java (hunk 3 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/category/CategoryController.java]: fixed

File: server/src/main/java/org/hyperskill/community/flashcards/card/CardService.java (hunk 1 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/CardService.java]: Thank you for noticing, you're absolutely right.

**Review by wisskirchenj — APPROVED** [review]

**Pull request merged**